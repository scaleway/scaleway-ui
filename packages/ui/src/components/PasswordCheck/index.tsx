import styled from '@emotion/styled'
import Icon from '../Icon'
import Stack from '../Stack'
import Text from '../Text'

type Rule = {
  name: string
  text: string
  valid: boolean
}

type PasswordCheckProps = {
  /**
   * Is an array of object that contains password rules. `name` is the name of the rule, `text` the text associated
   * with the rule and `valid` is a boolean that determine if the rule is respected or not.
   */
  rules: Rule[]
}

const PasswordCheckContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  gap: ${({ theme }) => theme.space['1']};
`

const PasswordCheck = ({ rules }: PasswordCheckProps) => (
  <PasswordCheckContainer>
    {rules.map(rule => (
      <Stack direction="row" gap={1} alignItems="center" key={rule.name}>
        <Icon
          name={rule.valid ? 'checkbox-circle-outline' : 'close-circle-outline'}
          color={rule.valid ? 'success' : 'neutral'}
          prominence="weak"
          size={20}
        />
        <Text as="p" variant="bodySmall">
          {rule.text}
        </Text>
      </Stack>
    ))}
  </PasswordCheckContainer>
)

export default PasswordCheck
