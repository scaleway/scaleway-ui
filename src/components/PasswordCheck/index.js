import { css } from '@emotion/react'
import PropTypes from 'prop-types'
import React from 'react'
import { Col, Grid, Row } from '../Grid'
import Icon from '../Icon'

const styles = {
  condition: css`
    display: flex;
    align-items: center;
    margin: 0 0 8px 0;
    font-size: 14px;
  `,
  icon: css`
    margin-right: 8px;
    align-self: center;
  `,
}

function PasswordCheck({ rules }) {
  return (
    <Grid fluid gutter={0}>
      <Row>
        {rules.map(rule => (
          <Col large={6} key={rule.name}>
            <p css={styles.condition}>
              <Icon
                name={
                  rule.valid ? 'check-circle-outline' : 'close-circle-outline'
                }
                color={rule.valid ? 'success' : 'darkGrey'}
                css={styles.icon}
                size={20}
              />
              {rule.text}
            </p>
          </Col>
        ))}
      </Row>
    </Grid>
  )
}

PasswordCheck.propTypes = {
  rules: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      valid: PropTypes.bool.isRequired,
    }),
  ).isRequired,
}

export default PasswordCheck
