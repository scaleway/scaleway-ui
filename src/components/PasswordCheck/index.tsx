import { css, useTheme } from '@emotion/react'
import PropTypes from 'prop-types'
import React, { FunctionComponent } from 'react'
import { Col, Grid, Row } from '../Grid'
import Icon from '../Icon'

const styles = {
  condition: css`
    display: flex;
    align-items: center;
    margin: 0 0 8px 0;
    font-size: 14px;
  `,
  icon: css`
    margin-right: 8px;
    align-self: center;
  `,
}

type Rule = {
  name: string
  text: string
  valid: boolean
}

type PasswordCheckProps = {
  /**
   * Is an array of object that contains password rules. `name` is the name of the rule, `text` the text associated
   * with the rule and `valid` is a boolean that determine if the rule is respected or not.
   */
  rules: Rule[]
}

const PasswordCheck: FunctionComponent<PasswordCheckProps> = ({ rules }) => {
  const theme = useTheme()

  return (
    <Grid fluid gutter={0}>
      <Row>
        {rules.map(rule => (
          <Col large={6} key={rule.name}>
            <p css={styles.condition}>
              <Icon
                name={
                  rule.valid ? 'check-circle-outline' : 'close-circle-outline'
                }
                color={
                  rule.valid
                    ? theme.colors.success.textWeak
                    : theme.colors.neutral.textWeak
                }
                css={styles.icon}
                size={20}
              />
              {rule.text}
            </p>
          </Col>
        ))}
      </Row>
    </Grid>
  )
}

PasswordCheck.propTypes = {
  rules: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      valid: PropTypes.bool.isRequired,
    }).isRequired,
  ).isRequired,
}

export default PasswordCheck
