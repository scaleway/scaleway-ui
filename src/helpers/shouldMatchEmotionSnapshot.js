/* eslint-disable import/no-extraneous-dependencies */
import createCache from '@emotion/cache'
import { createSerializer } from '@emotion/jest'
import { CacheProvider } from '@emotion/react'
import { render } from '@testing-library/react'
import React from 'react'

expect.addSnapshotSerializer(createSerializer())

const emotionCache = createCache({
  key: 'cache',
})
emotionCache.compat = true

export default (component, { transform } = {}) => {
  const node = render(
    <CacheProvider value={emotionCache}>{component}</CacheProvider>,
  )
  if (transform) transform(node)

  expect(node.asFragment()).toMatchSnapshot()
}
