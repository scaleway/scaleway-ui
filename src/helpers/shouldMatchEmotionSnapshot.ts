/* eslint-disable import/no-extraneous-dependencies */
import { createSerializer } from '@emotion/jest'
import { RenderOptions, render } from '@testing-library/react'
import { ReactNode } from 'react'
import defaultTheme from '../theme'
import renderWithTheme from './renderWithTheme'

// use only class hash (generated from css style content)
expect.addSnapshotSerializer(
  createSerializer({ classNameReplacer: className => className }),
)

interface Options {
  options?: RenderOptions
  theme?: typeof defaultTheme
  transform?: (node: ReturnType<typeof render>) => Promise<void> | void
}

export default async (
  component: ReactNode,
  { transform, options, theme }: Options = {},
): Promise<void> => {
  const node = renderWithTheme(component, options, theme)
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
  node.unmount()
}
