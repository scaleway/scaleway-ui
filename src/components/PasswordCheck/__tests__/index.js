import React from 'react'
import PasswordCheck from '..'
import shouldMatchEmotionSnapshot from '../../../helpers/shouldMatchEmotionSnapshot'

describe('PasswordCheck', () => {
  test('render with custom values', () => {
    shouldMatchEmotionSnapshot(
      <PasswordCheck
        rules={[
          {
            name: 'custom1',
            valid: false,
            text:
              "That's a beautiful custom password check we have right there",
          },
          {
            name: 'custom2',
            valid: true,
            text:
              "That's a second beautiful custom password check we have right there",
          },
        ]}
      />,
    )
  })
})
