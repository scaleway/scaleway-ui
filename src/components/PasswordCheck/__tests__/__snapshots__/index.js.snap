// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordCheck render with custom values 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  padding-left: 0;
  padding-right: 0;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -8px;
  margin-right: -8px;
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-left: 8px;
  padding-right: 8px;
}

@media (min-width: 992px) {
  .emotion-4 {
    -webkit-flex: 0 0 50%;
    -ms-flex: 0 0 50%;
    flex: 0 0 50%;
    max-width: 50%;
  }
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 0 8px 0;
  font-size: 14px;
}

.emotion-7 {
  fill: darkGrey;
  vertical-align: middle;
  height: 20px;
  width: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-right: 8px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.emotion-11 {
  fill: #45d6b5;
  vertical-align: middle;
  height: 20px;
  width: 20px;
  min-width: 20px;
  min-height: 20px;
  margin-right: 8px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<div
    class="sui-box emotion-0 emotion-1"
  >
    <div
      class="sui-box emotion-2 emotion-3"
    >
      <div
        class="sui-box emotion-4 emotion-5"
      >
        <p
          class="emotion-6"
        >
          <svg
            class="sui-box emotion-7"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2C6.47,2 2,6.47 2,12C2,17.53 6.47,22 12,22C17.53,22 22,17.53 22,12C22,6.47 17.53,2 12,2M14.59,8L12,10.59L9.41,8L8,9.41L10.59,12L8,14.59L9.41,16L12,13.41L14.59,16L16,14.59L13.41,12L16,9.41L14.59,8Z"
            />
          </svg>
          That's a beautiful custom password check we have right there
        </p>
      </div>
      <div
        class="sui-box emotion-4 emotion-5"
      >
        <p
          class="emotion-6"
        >
          <svg
            class="sui-box emotion-11"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4M11,16.5L6.5,12L7.91,10.59L11,13.67L16.59,8.09L18,9.5L11,16.5Z"
            />
          </svg>
          That's a second beautiful custom password check we have right there
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
